highfidelity/hifi
# CR Job #21281 -- Doppleganger Mirror

This PR adds support for using a "doppleganger mirror" to reflect your Avatar's model and joint data onto a private Overlay view.  It can be used to self-test various Avatar features like how your hand controllers are performing or animations look to others.  Unlike Camera-mode mirrors, to find the right vantage point you just walk or fly around your doppleganger using your normal movement controls.

#### Testing:
* Launch Interface
* Open the `Running Scripts...` dialog and load **system/app-doppleganger.js** from the tree view.
* Verify the script loads without errors and a new tablet app icon appears:
![image](https://cloud.githubusercontent.com/assets/94753/25407484/7467ad36-29d8-11e7-91e8-2dc151c1251b.png)
* Click the icon to activate the mirror.
* Verify the doppleganger appears in front of you and automatically mimics your joint movements.
* Click the icon again to disable the mirror.
* Verify the doppleganger disappears.

#### Exception Testing
(with **app-doppleganger.js** loaded)
* Avatar model changes:
  * Wear a valid avatar model and activate the mirror / doppleganger.
  * Change your avatar to a different model.
  * Verify the mirror automatically switches off.
* Bogus models / URLs:
  * Wear a valid avatar model and activate the mirror / doppleganger.
  * Switch your avatar model to a bogus URL (eg: https://httpbin.org/delay/5)
  * Verify that after a short delay (currently ~10 seconds) the mirror pops-up an error message and automatically disables itself.
* Domain changes
  * Wear a valid avatar model and activate the mirror / doppleganger.
  * Change to a different domain.
  * Verify the mirror automatically shuts itself off.

*note: Only joint data is synchronized in this version, so for example attachments and blendshapes will still only appear on your actual Avatar.*
